### PR TITLE
Simplified eliminators

### LINKED ISSUE #157: NonDetC can’t produce a Set in a single step
You have to do this inconvenient `runNonDet @[]` step first followed by `fmap mconcat`.

### PR DESCRIPTION
**Pull request by robrix**
This PR simplifies the `NonDetC` handler, and adds a couple of extra ones.

- [x] Fixes #157.
- [x] Depends on #196.
- [x] Depends on #197.
- [x] Depends on #198.
- [x] Depends on #199.
- [x] Depends on #200.
- [x] Depends on #204.
- [x] Depends on #205.
- [x] Depends on #219.

### CODE REVIEW

File: src/Control/Carrier/NonDet/Church.hs [hunk 1]
```diff
@@ -58,7 +66,7 @@ instance Alternative (NonDetC m) where
 
 instance Monad (NonDetC m) where
   NonDetC a >>= f = NonDetC $ \ fork leaf nil ->
-    a fork (\ a' -> runNonDetC (f a') fork leaf nil) nil
+    a fork (runNonDet fork leaf nil . f) nil
```
Comment (robrix) [inline src/Control/Carrier/NonDet/Church.hs]: This was by no means the point of this PR, but it’s certainly a nice side benefit that we can use the eliminators point-free.

File: src/Control/Carrier/Trace/Returning.hs [hunk 2]
```diff
@@ -5,7 +5,7 @@ module Control.Carrier.Trace.Returning
   -- * Trace carrier
 , runTrace
 , TraceC(..)
--- * Re-exports
```
Comment (robrix) [inline src/Control/Carrier/Trace/Returning.hs]: conform!

File: src/Control/Effect/Cull.hs [hunk 3]
```diff
@@ -36,5 +36,4 @@ cull m = send (Cull m pure)
 -- >>> import Test.QuickCheck
 -- >>> import Control.Carrier.Cull.Church
 -- >>> import Control.Carrier.NonDet.Church
--- >>> import Control.Carrier.Pure
```
Comment (robrix) [inline src/Control/Effect/Cull.hs]: This is re-exported by `Control.Carrier`, and I missed cleaning these up.

File: ChangeLog.md [hunk 4]
```diff
@@ -30,6 +30,12 @@
 
 - Removes the re-export of `Member` from all carrier modules, re-exporting `Has` in its place. `Has` constraints should generally be used instead, and specialist cases can import `Control.Effect.Sum` for `Member`. ([#217](https://github.com/fused-effects/fused-effects/pull/217))
 
+- Redesigns & renames the handlers for church-encoded nondeterminism carriers to standardize naming and usage patterns. ([#207](https://github.com/fused-effects/fused-effects/pull/207))
+  - The primary handlers (`runChoose`, `runNonDet`, `runCut`, `runCull`) take multiple continuations.
+  - Handlers which return an `Alternative` are suffixed with `A`, e.g. `runNonDetA`.
+  - Handlers which return a `Monoid` are suffixed with `M`, e.g. `runNonDetM`.
+  - Handlers which return a `Semigroup` are suffixed with `S`, e.g. `runChooseS`.
```
Comment (patrickt) [inline ChangeLog.md]: Loving this.

File: src/Control/Carrier/Choose/Church.hs [hunk 5]
```diff
@@ -20,11 +21,15 @@ import Control.Monad.Fix
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Class
 import Data.Maybe (fromJust)
+import qualified Data.Semigroup as S
 import Prelude hiding (fail)
 
 runChoose :: (m b -> m b -> m b) -> (a -> m b) -> ChooseC m a -> m b
 runChoose fork leaf m = runChooseC m fork leaf
 
+runChooseS :: (Applicative m, S.Semigroup b) => (a -> m b) -> ChooseC m a -> m b
```
Comment (patrickt) [inline src/Control/Carrier/Choose/Church.hs]: Can we get documentation for this, mentioning that it uses `<>` to merge its results?

File: src/Control/Carrier/Choose/Church.hs [hunk 6]
```diff
@@ -50,7 +55,9 @@ instance Fail.MonadFail m => Fail.MonadFail (ChooseC m) where
 
 instance MonadFix m => MonadFix (ChooseC m) where
   mfix f = ChooseC $ \ fork leaf ->
-    mfix (runChoose (liftA2 Fork) (pure . Leaf)
+    mfix (runChoose
+      (liftA2 Fork)
+      (pure . Leaf)
       . f . fromJust . fold (Alt.<|>) Just)
```
Comment (patrickt) [inline src/Control/Carrier/Choose/Church.hs]: Can we use something more descriptive than `fromJust` here? `fromMaybe (“error: ChooseC mfix returned Just”)`?

Comment (robrix) [reply src/Control/Carrier/Choose/Church.hs]: This is being replaced in #233 anyway, so I’mma leave it as-is.

**Pull request merged**